added PartialEq/ Eq for value::owned and added  unit tests for container types

desaikd opened the pull request:
*Issue #149 , #147*

*Description of changes:*
This PR works on creating `PartialEq` and `Eq` for `value::Owned` types and adds unit tests for container types.

*Changes:*
- Added `PartialEq` and `Eq` for `OwnedSymbolToken`, `OwnedImportSource`, `OwnedStruct`, `OwnedSequence`, `OwnedElement`. 
- Added unit tests for container types.

almann commented on src/value/owned.rs:
```diff
@@ -62,6 +74,24 @@ impl OwnedSymbolToken {
     }
 }
 
+impl PartialEq for OwnedSymbolToken {
+    fn eq(&self, other: &Self) -> bool {
+        if other.text != None && self.text != None {
+            if other.text == self.text {
+                return true;
+            }
+            false
+        } else {
+            if self.source == other.source {
+                return true;
+            }
+            false
+        }
+    }
+}
```
I might make this a bit more concise without the extra returns and explicit Boolean values

```rust
        if other.text != None && self.text != None {
            // both has text
            other.text == self.text
        } else {
            // one of the tokens has undefined text, the only way they can be equal if they have the same source
            // None implies $0
            self.source == other.source
        }
```

Note that we probably should document the `None` implies `$0` bit in our APIs (if you agree), and we really need tests for the different cases described in:

https://amzn.github.io/ion-docs/docs/symbols.html#data-model

desaikd replied:
Oh yes! I will change it to be more concise. And yes, I think we should document None implies $0. I will try to add tests for above mentioned cases.

almann commented on src/value/owned.rs:
```diff
@@ -234,6 +293,26 @@ impl OwnedElement {
     }
 }
 
+impl PartialEq for OwnedElement {
+    fn eq(&self, other: &Self) -> bool {
+        match &self.ion_type() {
+            IonType::Null => self.is_null() == other.is_null(),
+            IonType::Boolean => self.as_bool() == other.as_bool(),
+            IonType::Decimal => self.as_decimal() == other.as_decimal(),
+            IonType::Integer => self.as_any_int() == other.as_any_int(),
+            IonType::Float => self.as_f64() == other.as_f64(),
+            IonType::Timestamp => self.as_timestamp() == other.as_timestamp(),
+            IonType::Clob | IonType::Blob => self.as_bytes() == other.as_bytes(),
+            IonType::String => self.as_str() == other.as_str(),
+            IonType::Symbol => self.as_sym() == other.as_sym(),
+            IonType::Struct => self.as_struct() == other.as_struct(),
+            IonType::SExpression | IonType::List => self.as_sequence() == other.as_sequence(),
+        }
+    }
+}
```
Is it possible to lean on a `PartialEq` defined for `OwnedValue`?  Also don't you need to compare the annotations?  This needs its own explicit tests too.

almann reviewed (changes requested):
I would really like to see test cases for the non-trivial `PartialEq` methods, they are potentially error prone.

almann commented on src/value/owned.rs:
```diff
@@ -53,6 +49,7 @@ impl ImportSource for OwnedImportSource {
 }
 
 /// An owned implementation of [`SymbolToken`].
+/// Here text as `None` represents SID $0
```
This comment belongs on the trait definition (since it should be uniformly implemented that way).

almann reviewed (approved):
:rocket: LGTM.  Do you mind adding a GH issue to explicitly test symbol token on its own?

almann reviewed (changes requested):
I am not sure I grok the test factoring for annotation testing.  I would either factor that out as its own test or maybe explain to me how you're thinking about it.

desaikd commented on src/value/owned.rs:
```diff
@@ -600,6 +590,16 @@ mod value_tests {
             (AsStruct, &|e| assert_eq!(None, e.as_struct())),
         ];
+
+        // tests for element with annotations
+        let annotations = vec!["foo", "bar", "baz"];
+        let owned_elem = OwnedElement::new(
+            annotations.iter().map(|s| (*s).into()).collect(),
+            val.clone(),
+        );
+        let expected: Vec<String> = annotations.iter().map(|s| (*s).into()).collect();
+        assert_eq!(expected, annotation_strings(&owned_elem));
```
I think I was trying to add 3 annotation to  each value in test cases and test those 3 annotations as you mentioned. But I think it would make more sense to have separate unit test for this. I will try to see if value lists can be used here.

almann replied:
Here is a small, contrived example of mixing parameter cases and value lists.

```rust
#[cfg(test)]
mod test {
    use rstest::*;

    #[rstest]
    #[case::range_10_to_100(10, 100)]
    #[case::range_5_to_45(5, 45)]
    fn foo1(#[case] lval: i64, #[case] hval: i64, #[values(20, 30, 40)] other: i64) {
        assert!(lval < other);
        assert!(hval > other);
    }

    #[rstest(
        lval, hval,
        case::range_10_to_100(10, 100),
        case::range_5_to_45(5, 45),
        other => [20, 30, 40]
    )]
    fn foo2(lval: i64, hval: i64, other: i64) {
        assert!(lval < other);
        assert!(hval > other);
    }
}
```

The `foo2` example uses the macros I used, but apparently that is the [*old compact syntax*][1].  Apparently `rstest` 0.7 added this (my `ion-c-sys` tests used 0.6).  We don't need to port over now, but something for us to consider going forward.

[1]: https://docs.rs/rstest/0.7.0/rstest/attr.rstest.html#old-compact-syntax

almann replied:
I do think having equivalence and non-equivalence classes of annotations as a value list could be used in the value test to test `Eq` more comprehensively, FWIW.

almann replied:
See #158 (based on this PR) which makes some changes that illustrate what I was trying to articulate here.

almann commented:
This reminds me that we should also have these tests factored for the `borrowed` module as well.  Ideally, though not required, it would be nice to make the tests factored well for both (and possibly consider using [`rstest_reuse`][1] as needed to avoid copy pasting the parameters).

[1]: https://crates.io/crates/rstest_reuse

almann commented on src/value/owned.rs:
```diff
@@ -472,7 +546,30 @@ mod value_tests {
+            IonType::SExpression,
+            AsSequence,
+            &|e: &OwnedElement| {
+                assert_eq!(Some(&OwnedSequence::new(vec![OwnedElement::from(OwnedValue::Boolean(true)), OwnedElement::from(OwnedValue::Boolean(false))])), e.as_sequence());
```
IIRC, we have an `FromIterator` implementation that should make this a little more compact

almann commented on src/value/owned.rs:
```diff
@@ -472,7 +546,30 @@ mod value_tests {
+            }
+        ),
+        case::struct_(
+            OwnedValue::Struct(OwnedStruct::from_iter(vec![("greetings", OwnedElement::from(OwnedValue::String("hello".into())))].into_iter())),
+            IonType::Struct,
+            AsStruct,
+            &|e: &OwnedElement| {
+                assert_eq!(Some(&OwnedStruct::from_iter(vec![("greetings", OwnedElement::from(OwnedValue::String("hello".into())))].into_iter())), e.as_struct());
```
Likewise, see above comment.

almann reviewed (commented):
Minor factoring comments below.

desaikd commented:
*Changes with https://github.com/amzn/ion-rust/pull/152/commits/9e61da3022578df48ccf67e0dac49be331a391e7*
- Added `FromIterator` for Sequence 
- Modified tests for using `FromIterator` for `Struct` and `Sequence`

almann commented on src/value/owned.rs:
```diff
@@ -233,12 +244,7 @@ impl Struct for OwnedStruct {
-            if other_values != this_values {
+            if other.get_all(key).ne(value.iter().map(|(_s, v)| v)) {
                 return false;
             }
```
As I look at this I think some separate struct tests are in order, especially because I think we need to take `no_text_fields` into account.

Specifically, I wouldn't try to roll all of this testing into the value tests, I think they are complicated to have their own sets.

almann replied:
Also, what happens if the vector order is different for a given key, this technically doesn't work, does it?

In other words `{a: 1, a: 2}` should equal `{a: 2, a: 1}`, does that work?

almann replied:
I misread the code above--it should work (though tests needed), but it is an O(N**2) comparison, If we want to make this O(N lg N) or O(N), we need an intermediate `BTreeSet` or `HashSet`.

almann replied:
See #160.

almann commented on src/value/owned.rs:
```diff
@@ -547,27 +553,27 @@ mod value_tests {
         ),
         case::list(
-            OwnedValue::List(OwnedSequence::new(vec![OwnedElement::from(OwnedValue::Boolean(true)), OwnedElement::from(OwnedValue::Boolean(false))])),
+            OwnedValue::List(vec![OwnedValue::Boolean(true).into(), OwnedValue::Boolean(false).into()].into_iter().collect()),
```
Not sure if it is worth it with only two elements, but `into_iter().map(|v| v.into()).collect()` might be be a bit more concise.

almann reviewed (changes requested):
Can you verify that the semantics on `OwnedStruct`'s `PartialEq` works for unknown text symbols and duplicate key values out of order?

I don't think the tests are sufficient to verify this as written, but let me know if I missed something.

almann reviewed (approved):
Looks good contingent on fast follow up on #161.
